src/plugins/data_source_management/public/components/data_source_view/data_source_view.tsx

```diff
@@ -57,77 +63,140 @@ export class DataSourceView extends React.Component<DataSourceViewProps, DataSou
+      dataSourceId,
+      this.state.defaultDataSourceOption
+    );
+    handleDataSourceViewErrorWithSwitchToDefaultOption(handleDataSourceViewErrorParams);
+  }
+
+  async handleDataSourceLabelOrBackupWithDefaultOption(
+    dataSourceId: string,
+    defaultDataSource: string | null
+  ) {
+    const selectedDataSource = await this.getSelectedDataSource(dataSourceId, defaultDataSource);
+    if (!this._isMounted) return;
+    const filteredSelectedDataSourceOption = this.getFilteredDataSource(selectedDataSource);
+    // if the selectedOption has been filtered out, treat it as invalid id error and early return
+    if (filteredSelectedDataSourceOption) {
+      this.setState({
+        selectedOption: [filteredSelectedDataSourceOption],
+        defaultDataSource,
+      });
+      return;
+    }
+
+    await this.getAndSetDefaultDataSourceOption(dataSourceId, defaultDataSource);
+    this.handleInvalidDataSourceError(dataSourceId);
+  }
+
+  getDataSourceViewErrorParams(
+    dataSourceId: string,
+    filteredDefaultDataSourceOption: DataSourceOption | null
+  ) {
+    return {
+      changeState: this.onError.bind(this, filteredDefaultDataSourceOption),
+      notifications: this.props.notifications!,
+      failedDataSourceId: dataSourceId,
+      defaultDataSourceOption: filteredDefaultDataSourceOption,
+      handleSwitch: this.handleSwitchDefaultDatasource.bind(this, filteredDefaultDataSourceOption),
+      callback: this.props.onSelectedDataSources,
+    } as DataSourceViewErrorWithDefaultParams;
+  }
   async componentDidMount() {
     this._isMounted = true;
 
     const selectedOption = this.props.selectedOption;
     const option = selectedOption[0];
     const optionId = option.id;
-
     const defaultDataSource = this.props.uiSettings?.get('defaultDataSource', null) ?? null;
-    if (optionId === '' && !this.props.hideLocalCluster) {
-      this.setState({
-        selectedOption: [LocalCluster],
-        defaultDataSource,
-      });
-      if (this.props.onSelectedDataSources) {
-        this.props.onSelectedDataSources([LocalCluster]);
-      }
+
+    // 1. id and label both exist, set defaultDataSource directly and early return
+    if (this.isCompletDataSourceOption(option)) {
+      this.updateSelectedOptionState(defaultDataSource, option);
       return;
     }
-
-    if (
-      (optionId === '' && this.props.hideLocalCluster) ||
-      (this.props.dataSourceFilter &&
-        this.props.selectedOption.filter(this.props.dataSourceFilter).length === 0)
-    ) {
-      this.setState({
-        selectedOption: [],
-      });
-      if (this.props.onSelectedDataSources) {
-        this.props.onSelectedDataSources([]);
+    // 2. handle local cluster
+    if (option && optionId === '') {
+      // 2.1 handle legit
+      if (!this.props.hideLocalCluster) {
+        this.updateSelectedOptionState(defaultDataSource, LocalCluster);
+        return;
       }
+      this.clearSelectedOptionState();
```

Comment:
should this be same as handle invalid data source?

Reply:
what clearSelectedOptionState does:
1. clear the state for selectedOption
2. pass empty array to `onSelectedDataSources` if it exist
what handle invalid data source does:
1. show error state
2. set defaultDataSourceOption
3. show notification
4. pass empty array to `onSelectedDataSources` if it exist
5. show switch to default option if the defaultOption exist
the only overlap is `pass empty array to `onSelectedDataSources` if it exist`, therefore, I think they are different